## Challenge 1

For this project I have been asked to answer the following question: 

**"To what extent does the sentiment expressed in a tweet influence user engagement with the tweet (likes and retweets)?"**

### Programming challenge

Using your Twitter data collected via ```twarc```, the tasks I have to complete are: 

1. Determine which variables are relevant for the research question (including control variables that are relevant related to users and tweets).
2. Create a minimized dataset (dataset with only variables necessary to answer the research question).
3. Make sure that the minimized dataset is pseudonymized (identifying information about users are removed from user-related columns and from text).

I have to create two binary variables for the Twitter data based on the text column. They should be two **meaningful** categories for this data, and they should have either the value 0 (when the tweet is not part of that category) or 1 (when the tweet is in that category). 



* **Technical** summary:

**The following variables have been retained:**

created_at, id, full_text, retweet_count, favorite_count, followers_count, friends_count, screen_name.

**The following can be considered Control Variables: the time (Created at) the tweet has been created (if the tweet has been made during some environemntal scandal/emergency it may be considered with more sentiment expressed in it and consequent engagment), the Id, if the person who made the tweet was a famous journalist or politician or related to the scandal/emergency or somehow to such topic (the tweets query was on the word "Greenpeace"). Friends Count and Followers Count have been kept as Control Variables as well since the number of followers/friends may influence the engagment of the users with the tweets. The sentiment expressed in the Tweet is the main Independent Variable. Instead, the Retweet Count and the Favourite Count are the Dependent Variables since they are a measure of the engagment with the Tweet itself.**

### Step-by-step Interpretation:

**Importing Pandas**

In [1]:
import pandas as pd

**Importing the Dataset which is made of Tweets containing the word "Greenpeace"**

In [2]:
df_jsonl = pd.read_json('greenpeacetweets.jsonl', lines=True)

**Checking the n. of Rows the dataset contains**

In [3]:
len(df_jsonl) 

1197

**Displaying first 5 rows to have a preview of the Dataframe**

In [4]:
df_jsonl.head()

created_at                   id               id_str  \
0 2021-09-15 20:21:15+00:00  1438236491394785288  1438236491394785280   
1 2021-09-15 20:21:13+00:00  1438236482532151301  1438236482532151296   
2 2021-09-15 20:21:03+00:00  1438236443638370308  1438236443638370304   
3 2021-09-15 20:21:02+00:00  1438236436084391936  1438236436084391936   
4 2021-09-15 20:20:54+00:00  1438236402114772995  1438236402114772992   

                                           full_text  truncated  \
0  RT @GreenpeaceArg: - ¿Por qué tan elegante, Ho...      False   
1  ¡Excusas, excusas, excusas! Cómo las empresas ...      False   
2  RT @Herbert_Diess: Greenpeace müsste man erfin...      False   
3  RT @PBIcanada: We congratulate @GreenpeaceCA o...      False   
4  RT @cadebe_: @GJGamble @BorisJohnson @trussliz...      False   

  display_text_range                                           entities  \
0           [0, 117]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1           [0, 115]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2           [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3           [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4           [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  \
0  {'media': [{'id': 1438173629850832905, 'id_str...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                            metadata  \
0  {'iso_language_code': 'es', 'result_type': 're...   
1  {'iso_language_code': 'es', 'result_type': 're...   
2  {'iso_language_code': 'de', 'result_type': 're...   
3  {'iso_language_code': 'en', 'result_type': 're...   
4  {'iso_language_code': 'en', 'result_type': 're...   

                                              source  ...  is_quote_status  \
0  <a href="https://mobile.twitter.com" rel="nofo...  ...            False   
1  <a href="http://twitter.com/download/iphone" r...  ...            False   
2  <a href="http://twitter.com/download/android" ...  ...             True   
3  <a href="http://twitter.com/download/iphone" r...  ...            False   
4  <a href="http://twitter.com/download/android" ...  ...            False   

   retweet_count  favorite_count  favorited retweeted possibly_sensitive  \
0             14               0      False     False                0.0   
1              0               0      False     False                0.0   
2             41               0      False     False                NaN   
3              7               0      False     False                NaN   
4              1               0      False     False                NaN   

   lang  quoted_status_id quoted_status_id_str  quoted_status  
0    es               NaN                  NaN            NaN  
1    es               NaN                  NaN            NaN  
2    de      1.438075e+18         1.438075e+18            NaN  
3    en               NaN                  NaN            NaN  
4    en               NaN                  NaN            NaN  

[5 rows x 31 columns]

**Checking which columns are present in the Data**

In [5]:
df_jsonl.columns

Index(['created_at', 'id', 'id_str', 'full_text', 'truncated',
       'display_text_range', 'entities', 'extended_entities', 'metadata',
       'source', 'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str',
       'in_reply_to_screen_name', 'user', 'geo', 'coordinates', 'place',
       'contributors', 'retweeted_status', 'is_quote_status', 'retweet_count',
       'favorite_count', 'favorited', 'retweeted', 'possibly_sensitive',
       'lang', 'quoted_status_id', 'quoted_status_id_str', 'quoted_status'],
      dtype='object')

**Cheching the languages present in the Data**

In [6]:
df_jsonl[['lang']].value_counts()

lang
en      420
es      286
de      238
fr       80
pt       50
nl       44
und      31
it       15
pl       10
ca        5
cs        4
no        3
sv        2
th        2
et        1
in        1
ja        1
el        1
ro        1
ru        1
ar        1
dtype: int64

**Looking for missing values**

In [7]:
df_jsonl.isna().sum()

created_at                      0
id                              0
id_str                          0
full_text                       0
truncated                       0
display_text_range              0
entities                        0
extended_entities            1102
metadata                        0
source                          0
in_reply_to_status_id         982
in_reply_to_status_id_str     982
in_reply_to_user_id           973
in_reply_to_user_id_str       973
in_reply_to_screen_name       973
user                            0
geo                          1197
coordinates                  1197
place                        1195
contributors                 1197
retweeted_status              401
is_quote_status                 0
retweet_count                   0
favorite_count                  0
favorited                       0
retweeted                       0
possibly_sensitive            899
lang                            0
quoted_status_id             1031
quoted_status_

**Checking which in which values the Data have been stored**

In [8]:
df_jsonl.dtypes

created_at                   datetime64[ns, UTC]
id                                         int64
id_str                                     int64
full_text                                 object
truncated                                   bool
display_text_range                        object
entities                                  object
extended_entities                         object
metadata                                  object
source                                    object
in_reply_to_status_id                    float64
in_reply_to_status_id_str                float64
in_reply_to_user_id                      float64
in_reply_to_user_id_str                  float64
in_reply_to_screen_name                   object
user                                      object
geo                                      float64
coordinates                              float64
place                                     object
contributors                             float64
retweeted_status    

**Making a minimised Dataframe only with the variable relevant to answer my question (only in English language). Tasks 1 and 2 completed.**

In [9]:
en_df = df_jsonl[df_jsonl['lang']=='en'][['id','user','created_at','source','full_text','retweet_count','favorite_count']]

**Testing the new Dataset displaying the first 5 rows. As we can see the "User" column contain more dictionaries in it and Python is not able to vidualise it correctly**

In [10]:
en_df.head()

id                                               user  \
3   1438236436084391936  {'id': 27281240, 'id_str': '27281240', 'name':...   
4   1438236402114772995  {'id': 1119365287, 'id_str': '1119365287', 'na...   
6   1438236275283214343  {'id': 738106329080987648, 'id_str': '73810632...   
9   1438235849058045956  {'id': 20565828, 'id_str': '20565828', 'name':...   
10  1438235834415714308  {'id': 59791627, 'id_str': '59791627', 'name':...   

                  created_at  \
3  2021-09-15 20:21:02+00:00   
4  2021-09-15 20:20:54+00:00   
6  2021-09-15 20:20:23+00:00   
9  2021-09-15 20:18:42+00:00   
10 2021-09-15 20:18:38+00:00   

                                               source  \
3   <a href="http://twitter.com/download/iphone" r...   
4   <a href="http://twitter.com/download/android" ...   
6   <a href="https://mobile.twitter.com" rel="nofo...   
9   <a href="http://twitter.com/download/iphone" r...   
10  <a href="http://twitter.com/download/android" ...   

                                            full_text  retweet_count  \
3   RT @PBIcanada: We congratulate @GreenpeaceCA o...              7   
4   RT @cadebe_: @GJGamble @BorisJohnson @trussliz...              1   
6   RT @GreenpeaceCA: I scream, you scream, WE ALL...              2   
9   RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...             16   
10  RT @zoev213: Happy 50th @greenpeaceusa Join me...              2   

    favorite_count  
3                0  
4                0  
6                0  
9                0  
10               0

**Checking lenght of the new Dataset**

In [11]:
len(en_df)

420

In [12]:
en_df[['retweet_count','favorite_count']].describe()

retweet_count  favorite_count
count     420.000000      420.000000
mean       24.057143        1.890476
std        57.408028       29.456613
min         0.000000        0.000000
25%         0.000000        0.000000
50%         4.000000        0.000000
75%        25.000000        0.000000
max       648.000000      602.000000

**Creating a minimised Dataframe for the Users. With this command I am asking Pandas to flatten the nested dictionaries and  converts them into separate columns** 

In [13]:
users = pd.json_normalize(en_df['user'])

**Testing if the command worked. It did**

In [14]:
users

id              id_str                        name  \
0              27281240            27281240                  wei wu wei   
1            1119365287          1119365287                      Elize⚘   
2    738106329080987648  738106329080987648            Craig Laferriere   
3              20565828            20565828  CitizenWonk #DCStatehood 💉   
4              59791627            59791627               Mark Trudgeon   
..                  ...                 ...                         ...   
415          1022655642          1022655642    Backdoor Conquistador 🇬🇷   
416          2925571294          2925571294                     emgfind   
417  882766377660166144  882766377660166144             Adrienne Moreau   
418  792280784468111360  792280784468111360             Innocent Indeje   
419            23247549            23247549                      johnny   

         screen_name               location  \
0    Osmotheque78000        California, USA   
1       ElizeCronje1                Big Bay   
2     JackLaferriere   Mississauga, Ontario   
3        CitizenWonk  Washington, DC USA 🇺🇸   
4      mark_trudgeon            Southampton   
..               ...                    ...   
415     LiteralDiego                          
416          emgfind        Alberta, Canada   
417       AdriMoreau          Dark dungeons   
418       ian_indeje                  Kenya   
419    redwingjohnny       windsor, ontario   

                                           description  \
0                Buddhist - Daoist - Quaker Questioner   
1    I am a happy postive person and don't get angr...   
2    Craig Laferrière lives in Mississauga, Ontario...   
3    anti-fascist Political Scientist #ONEV1 Z22. R...   
4    Green Supporter and campaigner, Saints season ...   
..                                                 ...   
415                                  Just 🐝-ing myself   
416  interests:exposed environmentalconcerns&coveru...   
417  Cat. Co-writer of brutal fiction and "fan-fict...   
418  SRHR Advocate & TOT, Peer Educator, teacher by...   
419  Just a city boy....... Born and raised in Sout...   

                         url  protected  followers_count  friends_count  ...  \
0                       None      False             5665           6235  ...   
1                       None      False             1165           1722  ...   
2    https://t.co/r2Bw6rEwjo      False               25            116  ...   
3                       None      False            32259          27188  ...   
4                       None      False              340           1516  ...   
..                       ...        ...              ...            ...  ...   
415                     None      False               35            386  ...   
416                     None      False             4663           3828  ...   
417                     None      False             1521           1385  ...   
418                     None      False             1521           1103  ...   
419  https://t.co/W5ZMWdIvCM      False             4034           3869  ...   

     default_profile default_profile_image  following follow_request_sent  \
0              False                 False      False               False   
1               True                 False      False               False   
2               True                 False      False               False   
3              False                 False      False               False   
4              False                 False      False               False   
..               ...                   ...        ...                 ...   
415             True                 False      False               False   
416             True                 False      False               False   
417             True                 False      False               False   
418             True                 False      False               False   
419            False                 False      Fa

**Checking the columns contained in the nested dictionaries** 

In [15]:
users.columns

Index(['id', 'id_str', 'name', 'screen_name', 'location', 'description', 'url',
       'protected', 'followers_count', 'friends_count', 'listed_count',
       'created_at', 'favourites_count', 'utc_offset', 'time_zone',
       'geo_enabled', 'verified', 'statuses_count', 'lang',
       'contributors_enabled', 'is_translator', 'is_translation_enabled',
       'profile_background_color', 'profile_background_image_url',
       'profile_background_image_url_https', 'profile_background_tile',
       'profile_image_url', 'profile_image_url_https', 'profile_link_color',
       'profile_sidebar_border_color', 'profile_sidebar_fill_color',
       'profile_text_color', 'profile_use_background_image',
       'has_extended_profile', 'default_profile', 'default_profile_image',
       'following', 'follow_request_sent', 'notifications', 'translator_type',
       'withheld_in_countries', 'entities.description.urls',
       'profile_banner_url', 'entities.url.urls'],
      dtype='object')

**Checking for missing values in the new minimised Dataset**

In [16]:
users.isna().sum()

id                                      0
id_str                                  0
name                                    0
screen_name                             0
location                                0
description                             0
url                                   262
protected                               0
followers_count                         0
friends_count                           0
listed_count                            0
created_at                              0
favourites_count                        0
utc_offset                            420
time_zone                             420
geo_enabled                             0
verified                                0
statuses_count                          0
lang                                  420
contributors_enabled                    0
is_translator                           0
is_translation_enabled                  0
profile_background_color                0
profile_background_image_url      

**Joining my 2 minimised Datasets together**

In [17]:
tweets = en_df.join(users, rsuffix = '_user')

**As we can see lenght hasn't changed**

In [18]:
len(tweets)

420

**All the columns have been categorised together**

In [19]:
tweets.columns

Index(['id', 'user', 'created_at', 'source', 'full_text', 'retweet_count',
       'favorite_count', 'id_user', 'id_str', 'name', 'screen_name',
       'location', 'description', 'url', 'protected', 'followers_count',
       'friends_count', 'listed_count', 'created_at_user', 'favourites_count',
       'utc_offset', 'time_zone', 'geo_enabled', 'verified', 'statuses_count',
       'lang', 'contributors_enabled', 'is_translator',
       'is_translation_enabled', 'profile_background_color',
       'profile_background_image_url', 'profile_background_image_url_https',
       'profile_background_tile', 'profile_image_url',
       'profile_image_url_https', 'profile_link_color',
       'profile_sidebar_border_color', 'profile_sidebar_fill_color',
       'profile_text_color', 'profile_use_background_image',
       'has_extended_profile', 'default_profile', 'default_profile_image',
       'following', 'follow_request_sent', 'notifications', 'translator_type',
       'withheld_in_countries', 'ent

In [20]:
tweets.isna().sum()

id                                      0
user                                    0
created_at                              0
source                                  0
full_text                               0
retweet_count                           0
favorite_count                          0
id_user                               252
id_str                                252
name                                  252
screen_name                           252
location                              252
description                           252
url                                   351
protected                             252
followers_count                       252
friends_count                         252
listed_count                          252
created_at_user                       252
favourites_count                      252
utc_offset                            420
time_zone                             420
geo_enabled                           252
verified                          

**Creating my final minimised dataframe with all the infi I need to answer my question**

In [21]:
final_df = tweets[['id','screen_name','created_at','source','full_text','retweet_count','favorite_count','followers_count','friends_count']]

In [22]:
final_df

id    screen_name                created_at  \
3     1438236436084391936    CitizenWonk 2021-09-15 20:21:02+00:00   
4     1438236402114772995  mark_trudgeon 2021-09-15 20:20:54+00:00   
6     1438236275283214343     AspieandMe 2021-09-15 20:20:23+00:00   
9     1438235849058045956      gazetka75 2021-09-15 20:18:42+00:00   
10    1438235834415714308      arynpekik 2021-09-15 20:18:38+00:00   
...                   ...            ...                       ...   
1167  1438192539400429577            NaN 2021-09-15 17:26:36+00:00   
1168  1438192531787825157            NaN 2021-09-15 17:26:34+00:00   
1177  1438192131206782979            NaN 2021-09-15 17:24:59+00:00   
1183  1438192017847226369            NaN 2021-09-15 17:24:32+00:00   
1186  1438192006350651396            NaN 2021-09-15 17:24:29+00:00   

                                                 source  \
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/android" ...   
6     <a href="https://mobile.twitter.com" rel="nofo...   
9     <a href="http://twitter.com/download/iphone" r...   
10    <a href="http://twitter.com/download/android" ...   
...                                                 ...   
1167  <a href="http://twitter.com/download/iphone" r...   
1168  <a href="http://twitter.com/download/android" ...   
1177  <a href="https://mobile.twitter.com" rel="nofo...   
1183  <a href="http://twitter.com/download/android" ...   
1186  <a href="https://mobile.twitter.com" rel="nofo...   

                                              full_text  retweet_count  \
3     RT @PBIcanada: We congratulate @GreenpeaceCA o...              7   
4     RT @cadebe_: @GJGamble @BorisJohnson @trussliz...              1   
6     RT @GreenpeaceCA: I scream, you scream, WE ALL...              2   
9     RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...             16   
10    RT @zoev213: Happy 50th @greenpeaceusa Join me...              2   
...                                                 ...            ...   
1167  @cowboy_satin @23brookside @People4Bernie @Ber...              0   
1168  RT @jubileam8: #Glyphosate spraying purposeful...              5   
1177  RT @EDProgramme: Companies like ExxonMobil kne...             17   
1183  RT @AbukaAlfred: Different countries, differen...              5   
1186    happy 50th Greenpeace ! https://t.co/gtBWdYOGRI              0   

      favorite_count  followers_count  friends_count  
3                  0          32259.0        27188.0  
4                  0            340.0         1516.0  
6                  0            387.0          672.0  
9                  0              1.0           35.0  
10                 0            446.0          443.0  
...              ...              ...            ...  
1167               1              NaN            NaN  
1168               0              NaN            NaN  
1177               0              NaN            NaN  
1183               0              NaN            NaN  
1186               0              NaN            NaN  

[420 rows x 9 columns]

**As we can see there are 252 missing values in 2 of my columns**

In [23]:
final_df.isna().sum()

id                   0
screen_name        252
created_at           0
source               0
full_text            0
retweet_count        0
favorite_count       0
followers_count    252
friends_count      252
dtype: int64

In [24]:
final_df[['friends_count','followers_count']].describe()

friends_count  followers_count
count     168.000000     1.680000e+02
mean     1837.047619     1.522012e+04
std      3291.624330     1.465437e+05
min         1.000000     0.000000e+00
25%       264.000000     1.932500e+02
50%       893.000000     8.600000e+02
75%      1990.500000     2.084000e+03
max     27188.000000     1.889553e+06

In [25]:
pd.set_option('display.float_format', lambda x: '%.5f' % x)

In [26]:
final_df[['friends_count','followers_count']].describe()

friends_count  followers_count
count      168.00000        168.00000
mean      1837.04762      15220.11905
std       3291.62433     146543.71983
min          1.00000          0.00000
25%        264.00000        193.25000
50%        893.00000        860.00000
75%       1990.50000       2084.00000
max      27188.00000    1889553.00000

**I have been asked to replace the missing values with a 0 and this is what I am doing in the next operations**

In [27]:
final_df['friendsCount_no_na'] = final_df['friends_count'].fillna(0)

<ipython-input-27-f93d8df9069a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df['friendsCount_no_na'] = final_df['friends_count'].fillna(0)


In [28]:
final_df['followers_count_no_na'] = final_df['followers_count'].fillna(0)

<ipython-input-28-04b11e68e653>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df['followers_count_no_na'] = final_df['followers_count'].fillna(0)


**Slicing the Dataframe excluding the columns with missing values (they have been replaced by 2 new columns for which the missing values have been replaced by a zero)**

In [29]:
final_df2 = final_df[['id','screen_name','created_at','source','full_text','retweet_count','favorite_count','followers_count_no_na','friendsCount_no_na']]

In [30]:
final_df2.isna().sum()

id                         0
screen_name              252
created_at                 0
source                     0
full_text                  0
retweet_count              0
favorite_count             0
followers_count_no_na      0
friendsCount_no_na         0
dtype: int64

In [31]:
final_df2.dtypes

id                                     int64
screen_name                           object
created_at               datetime64[ns, UTC]
source                                object
full_text                             object
retweet_count                          int64
favorite_count                         int64
followers_count_no_na                float64
friendsCount_no_na                   float64
dtype: object

**Now let's be sure that the minimized dataset is pseudonymized (identifying information about users are removed from user-related columns and from text). Unique users have been picked up asking pandas to drop the duplicates (In case the same user posted more then one Tweet)**

In [32]:
users = tweets[['screen_name']].drop_duplicates()

In [33]:
users.head()

screen_name
3     CitizenWonk
4   mark_trudgeon
6      AspieandMe
9       gazetka75
10      arynpekik

**Here the index column has been resettled and matched with a newly created pseudID used to re-name the index coloumn**

In [34]:
users = users.reset_index()

**The pseudID has been matched with the "screen name" of the user through the creation of a dictionary which contains every pseuID corrisponding at each screen name**

In [35]:
users = users.rename(columns={'index': 'pseudID'})

**A function to have all the sceen name in lower case has been created in order to have pandas able to find all the screen name in the dictionary. Task 3 completed**

In [36]:
userids = {}
for screen_name, pseudID in users[['screen_name', 'pseudID']].values.tolist():
    screen_name = str(screen_name).lower()
    userids[screen_name] = str(pseudID).lower()

In [37]:
def pseudonimise_user(row):
    row['pseudID'] = userids[str(row['screen_name']).lower()]
    return row

In [38]:
df_ps = final_df2.apply(pseudonimise_user, axis=1)

In [39]:
del df_ps['screen_name']

In [40]:
df_pseudo= df_ps.replace(to_replace ='@\S+', value = '@mention', regex = True)

In [41]:
df_pseudo

id                created_at  \
3     1438236436084391936 2021-09-15 20:21:02+00:00   
4     1438236402114772995 2021-09-15 20:20:54+00:00   
6     1438236275283214343 2021-09-15 20:20:23+00:00   
9     1438235849058045956 2021-09-15 20:18:42+00:00   
10    1438235834415714308 2021-09-15 20:18:38+00:00   
...                   ...                       ...   
1167  1438192539400429577 2021-09-15 17:26:36+00:00   
1168  1438192531787825157 2021-09-15 17:26:34+00:00   
1177  1438192131206782979 2021-09-15 17:24:59+00:00   
1183  1438192017847226369 2021-09-15 17:24:32+00:00   
1186  1438192006350651396 2021-09-15 17:24:29+00:00   

                                                 source  \
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/android" ...   
6     <a href="https://mobile.twitter.com" rel="nofo...   
9     <a href="http://twitter.com/download/iphone" r...   
10    <a href="http://twitter.com/download/android" ...   
...                                                 ...   
1167  <a href="http://twitter.com/download/iphone" r...   
1168  <a href="http://twitter.com/download/android" ...   
1177  <a href="https://mobile.twitter.com" rel="nofo...   
1183  <a href="http://twitter.com/download/android" ...   
1186  <a href="https://mobile.twitter.com" rel="nofo...   

                                              full_text  retweet_count  \
3     RT @mention We congratulate @mention on their ...              7   
4     RT @mention @mention @mention @mention @mentio...              1   
6     RT @mention I scream, you scream, WE ALL SCREA...              2   
9     RT @mention TODAY IN HISTORY\n\n1971–Environme...             16   
10    RT @mention Happy 50th @mention Join me and su...              2   
...                                                 ...            ...   
1167  @mention @mention @mention @mention @mention @...              0   
1168  RT @mention #Glyphosate spraying purposefully ...              5   
1177  RT @mention Companies like ExxonMobil knew abo...             17   
1183  RT @mention Different countries, different rac...              5   
1186    happy 50th Greenpeace ! https://t.co/gtBWdYOGRI              0   

      favorite_count  followers_count_no_na  friendsCount_no_na pseudID  
3                  0            32259.00000         27188.00000       3  
4                  0              340.00000          1516.00000       4  
6                  0              387.00000           672.00000       6  
9                  0                1.00000            35.00000       9  
10                 0              446.00000           443.00000      10  
...              ...                    ...                 ...     ...  
1167               1                0.00000             0.00000     421  
1168               0                0.00000             0.00000     421  
1177               0                0.00000             0.00000     421  
1183               0                0.00000             0.00000     421  
1186               0                0.00000             0.00000     421  

[420 rows x 9 columns]

## Challenge 2

For this challenge I will need to merge the sentiment analysis results with the Dataset and check whether the length of the dataframe generated by the merge makes sense.

### Step-by-step Interpretation:



**Importing my Sentiment Analysis**

In [42]:
sent = pd.read_pickle('FlorindoArgondizzo_EN_completed.pkl')

**Checking the lenght of the Data**

In [43]:
len(sent)

420

**Exploring the Dataframe**

In [44]:
sent

id positive negative neutral
3     1438236436084391936        3       -1       1
4     1438236402114772995        1       -1       0
6     1438236275283214343        2       -5      -1
9     1438235849058045956        2       -1       1
10    1438235834415714308        3       -1       1
...                   ...      ...      ...     ...
1167  1438192539400429577        2       -1       1
1168  1438192531787825157        1       -3      -1
1177  1438192131206782979        2       -1       1
1183  1438192017847226369        1       -2      -1
1186  1438192006350651396        3       -1       1

[420 rows x 4 columns]

In [45]:
sent.columns

Index(['id', 'positive', 'negative', 'neutral'], dtype='object')

In [46]:
sent.isna().sum()

id          0
positive    0
negative    0
neutral     0
dtype: int64

In [47]:
sent.dtypes

id          object
positive    object
negative    object
neutral     object
dtype: object

**Changing the type of value on which the Sentiment Analysis results has been stored especially since I will use the "id" column as unique identifier for the merging, consequently, the type of data on which the merge will be made must be the same (both integers in this case)**

In [48]:
sent[['id','positive','negative','neutral']] = sent[['id','positive','negative','neutral']].apply(pd.to_numeric)

In [49]:
sent.dtypes

id          int64
positive    int64
negative    int64
neutral     int64
dtype: object

**Making an inner merging between the 2 Dataframes using the "id" column as unique identifier**

In [50]:
complete = final_df2.merge(sent, on='id', how='inner')

**Checking the Dataframe**

In [51]:
complete

id    screen_name                created_at  \
0    1438236436084391936    CitizenWonk 2021-09-15 20:21:02+00:00   
1    1438236402114772995  mark_trudgeon 2021-09-15 20:20:54+00:00   
2    1438236275283214343     AspieandMe 2021-09-15 20:20:23+00:00   
3    1438235849058045956      gazetka75 2021-09-15 20:18:42+00:00   
4    1438235834415714308      arynpekik 2021-09-15 20:18:38+00:00   
..                   ...            ...                       ...   
415  1438192539400429577            NaN 2021-09-15 17:26:36+00:00   
416  1438192531787825157            NaN 2021-09-15 17:26:34+00:00   
417  1438192131206782979            NaN 2021-09-15 17:24:59+00:00   
418  1438192017847226369            NaN 2021-09-15 17:24:32+00:00   
419  1438192006350651396            NaN 2021-09-15 17:24:29+00:00   

                                                source  \
0    <a href="http://twitter.com/download/iphone" r...   
1    <a href="http://twitter.com/download/android" ...   
2    <a href="https://mobile.twitter.com" rel="nofo...   
3    <a href="http://twitter.com/download/iphone" r...   
4    <a href="http://twitter.com/download/android" ...   
..                                                 ...   
415  <a href="http://twitter.com/download/iphone" r...   
416  <a href="http://twitter.com/download/android" ...   
417  <a href="https://mobile.twitter.com" rel="nofo...   
418  <a href="http://twitter.com/download/android" ...   
419  <a href="https://mobile.twitter.com" rel="nofo...   

                                             full_text  retweet_count  \
0    RT @PBIcanada: We congratulate @GreenpeaceCA o...              7   
1    RT @cadebe_: @GJGamble @BorisJohnson @trussliz...              1   
2    RT @GreenpeaceCA: I scream, you scream, WE ALL...              2   
3    RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...             16   
4    RT @zoev213: Happy 50th @greenpeaceusa Join me...              2   
..                                                 ...            ...   
415  @cowboy_satin @23brookside @People4Bernie @Ber...              0   
416  RT @jubileam8: #Glyphosate spraying purposeful...              5   
417  RT @EDProgramme: Companies like ExxonMobil kne...             17   
418  RT @AbukaAlfred: Different countries, differen...              5   
419    happy 50th Greenpeace ! https://t.co/gtBWdYOGRI              0   

     favorite_count  followers_count_no_na  friendsCount_no_na  positive  \
0                 0            32259.00000         27188.00000         3   
1                 0              340.00000          1516.00000         1   
2                 0              387.00000           672.00000         2   
3                 0                1.00000            35.00000         2   
4                 0              446.00000           443.00000         3   
..              ...                    ...                 ...       ...   
415               1                0.00000             0.00000         2   
416               0                0.00000             0.00000         1   
417               0                0.00000             0.00000         2   
418               0                0.00000             0.00000         1   
419               0                0.00000             0.00000         3   

     negative  neutral  
0          -1        1  
1          -1        0  
2          -5       -1  
3          -1        1  
4          -1        1  
..        ...      ...  
415        -1        1  
416        -3       -1  
417        -1        1  
418        -2       -1  
419        -1        1  

[420 rows x 12 columns]

In [52]:
complete.columns

Index(['id', 'screen_name', 'created_at', 'source', 'full_text',
       'retweet_count', 'favorite_count', 'followers_count_no_na',
       'friendsCount_no_na', 'positive', 'negative', 'neutral'],
      dtype='object')

In [53]:
complete.isna().sum()

id                         0
screen_name              252
created_at                 0
source                     0
full_text                  0
retweet_count              0
favorite_count             0
followers_count_no_na      0
friendsCount_no_na         0
positive                   0
negative                   0
neutral                    0
dtype: int64

## Challenge 3

The results of the sentiment analysis present three different columns: ```neutral```,  ```positive```, and ```negative```.

For this challenge, I will need to:
1. Create one variable that summarizes the sentiment (i.e., that somehow aggregates the information of it being positive or negative - or potentially neutral - into one single variable)
2. Describe the sentiment of your tweets (mean, SD, mode - select metrics that make sense depending on how you created the sentiment variable).
3. Create a new dataframe taking a random sample of 15 tweets from your dataset. 


### Step-by-step Interpretation:


**Exploring in details the results of the Sentiment Analysis**

In [54]:
complete['positive'].value_counts()

1    188
2    143
3     81
4      8
Name: positive, dtype: int64

In [55]:
complete['negative'].value_counts()

-1    286
-2     72
-3     41
-4     14
-5      7
Name: negative, dtype: int64

In [56]:
complete['neutral'].value_counts()

 1    190
-1    124
 0    106
Name: neutral, dtype: int64

**Summarizing the scores creating a new column with the sum of the negative and the positive scores together (the neutral scores will be ignored)**

In [57]:
complete['sent_score'] = complete['negative'] + complete['positive']

In [58]:
complete

id    screen_name                created_at  \
0    1438236436084391936    CitizenWonk 2021-09-15 20:21:02+00:00   
1    1438236402114772995  mark_trudgeon 2021-09-15 20:20:54+00:00   
2    1438236275283214343     AspieandMe 2021-09-15 20:20:23+00:00   
3    1438235849058045956      gazetka75 2021-09-15 20:18:42+00:00   
4    1438235834415714308      arynpekik 2021-09-15 20:18:38+00:00   
..                   ...            ...                       ...   
415  1438192539400429577            NaN 2021-09-15 17:26:36+00:00   
416  1438192531787825157            NaN 2021-09-15 17:26:34+00:00   
417  1438192131206782979            NaN 2021-09-15 17:24:59+00:00   
418  1438192017847226369            NaN 2021-09-15 17:24:32+00:00   
419  1438192006350651396            NaN 2021-09-15 17:24:29+00:00   

                                                source  \
0    <a href="http://twitter.com/download/iphone" r...   
1    <a href="http://twitter.com/download/android" ...   
2    <a href="https://mobile.twitter.com" rel="nofo...   
3    <a href="http://twitter.com/download/iphone" r...   
4    <a href="http://twitter.com/download/android" ...   
..                                                 ...   
415  <a href="http://twitter.com/download/iphone" r...   
416  <a href="http://twitter.com/download/android" ...   
417  <a href="https://mobile.twitter.com" rel="nofo...   
418  <a href="http://twitter.com/download/android" ...   
419  <a href="https://mobile.twitter.com" rel="nofo...   

                                             full_text  retweet_count  \
0    RT @PBIcanada: We congratulate @GreenpeaceCA o...              7   
1    RT @cadebe_: @GJGamble @BorisJohnson @trussliz...              1   
2    RT @GreenpeaceCA: I scream, you scream, WE ALL...              2   
3    RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...             16   
4    RT @zoev213: Happy 50th @greenpeaceusa Join me...              2   
..                                                 ...            ...   
415  @cowboy_satin @23brookside @People4Bernie @Ber...              0   
416  RT @jubileam8: #Glyphosate spraying purposeful...              5   
417  RT @EDProgramme: Companies like ExxonMobil kne...             17   
418  RT @AbukaAlfred: Different countries, differen...              5   
419    happy 50th Greenpeace ! https://t.co/gtBWdYOGRI              0   

     favorite_count  followers_count_no_na  friendsCount_no_na  positive  \
0                 0            32259.00000         27188.00000         3   
1                 0              340.00000          1516.00000         1   
2                 0              387.00000           672.00000         2   
3                 0                1.00000            35.00000         2   
4                 0              446.00000           443.00000         3   
..              ...                    ...                 ...       ...   
415               1                0.00000             0.00000         2   
416               0                0.00000             0.00000         1   
417               0                0.00000             0.00000         2   
418               0                0.00000             0.00000         1   
419               0                0.00000             0.00000         3   

     negative  neutral  sent_score  
0          -1        1           2  
1          -1        0           0  
2          -5       -1          -3  
3          -1        1           1  
4          -1        1           2  
..        ...      ...         ...  
415        -1        1           1  
416        -3       -1          -2  
417        -1        1           1  
418        -2       -1          -1  
419        -1        1           2  

[420 rows x 13 columns]

**Asking Pandas to show me up to 500 rows of the Dataframe**

In [59]:
pd.set_option('display.max_rows', 500)

In [60]:
complete[['full_text','sent_score']]

full_text  sent_score
0    RT @PBIcanada: We congratulate @GreenpeaceCA o...           2
1    RT @cadebe_: @GJGamble @BorisJohnson @trussliz...           0
2    RT @GreenpeaceCA: I scream, you scream, WE ALL...          -3
3    RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...           1
4    RT @zoev213: Happy 50th @greenpeaceusa Join me...           2
5    RT @Greenpeace: "Today we need huge, unprecede...           0
6    Im fundraising  a speed climb supporting Green...           1
7    RT @cosmicadriestar: HAPPY BIRTHDAY GREENPEACE!!!           2
8    @Greenpeace @climatemorgan Dolphins from the F...           2
9    @AllatRaTV @BardemAntarctic @carlosbardem @gre...          -1
10   RT @AgBioWorld: “And what drove this wicked ca...           2
11   RT @Greenpeace: Our planet is worth more than ...           1
12   RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...           1
13   RT @zoev213: Happy 50th @greenpeaceusa Join me...           2
14   RT @Greenpeace: Our planet is worth more than ...           1
15   @Greenpeace Let’s instead sue the renewables i...          -1
16   RT @MarineGOfficial: Dear Twitter,\nMy name is...           1
17   RT @AllatRaTV: Mr. #JavierBardem, today #Globa...           0
18   @AllatRaTV @BardemAntarctic @carlosbardem @gre...          -1
19   RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...           1
20   A report this week from environmental group @G...          -1
21   RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...           1
22   RT @TheQuokkaCo: It’s Greenpeace Day!  Do you ...           1
23   @DeborahMeaden @BarbaraSutton15 @Tesco I don't...           0
24   Happy 50th @greenpeaceusa Join me and support ...           2
25   RT @David_Ritter: 50 years ago today, Greenpea...          -1
26   RT @Christina_T81: The Rainbow Warrior just ar...           0
27   RT @reggi_124: @Katie_Malarkey @candacekellyto...           0
28   RT @MarineGOfficial: Dear Twitter,\nMy name is...           1
29   RT @greenpeaceusa: Today is Greenpeace's 50th ...           1
30   RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...           1
31   Amazing, look at what we can do 👌🏻🐝🌿this is ju...           3
32   RT @Luisamneubauer: Happy Birthday Greenpeace!...           2
33   @Greenpeace RAW SEWAGE DUMPING IN FRESH WATERS...           0
34   RT @Greenpeace: Our planet is worth more than ...           1
35   RT @youmustdo1: #BREAKING \nI always said that...           0
36   RT @youmustdo1: #BREAKING \nI always said that...           0
37   RT @ShaneMiley3: https://t.co/MUL6zOcc2E\n\nPl...           1
38   RT @youmustdo1: #BREAKING \nI always said that...           0
39   RT @T_Pricey: @TonyHWindsor A million dollars ...          -1
40   .@Ebony_4_Justice has become the 1st Black per...           1
41   RT @Greenpeace: "Today we need huge, unprecede...           0
42   RT @RogerHallamCS21: UK trailing most European...           0
43   RT @CoreyBrotherson: It’s about time to previe...           0
44   @skhalymon @BardemAntarctic @carlosbardem @gre...          -1
45   an investigation by @Greenpeace found that the...           0
46   RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...           1
47   @james1664 @GJGamble @BorisJohnson @trussliz @...          -1
48   (March 12, 2019) Former Greenpeace founder Pat...          -1
49   RT @skhalymon: Dear  #JavierBardem!  Please ac...           2
50   RT @enckj: Great news from @PepsiCo esp if the...           2
51   Vermisse die Statements von #wwf #nabu #Greenp...          -3
52   RT @greenpeaceusa: Today is Greenpeace's 50th ...           1
53   RT @greenpeaceusa: The @USChamber trying to ke...           0
54   @LBC the start of this video the protesters ar...          -1
55   Happy birthday @Greenpeace 💚🌱🌍 I love you, tha...           3
56   RT @Greenpeace: "Today we need huge, unprecede...           0
57   RT @danielmittler: Happy 50th birthday @Greenp...           1
58   @AdamBlazowski @PurcellGregory @AnthonyBriggs ...           0
59   RT @LongTimeAmy: TOD

**Checking final results of the sum of the 2 scores (negative and positive)**

In [61]:
complete['sent_score'].value_counts()

 0    124
 1    119
 2     64
-1     64
-2     29
-3     11
 3      7
-4      2
Name: sent_score, dtype: int64

**Asking Python to show me the results in form of percentages: as we can see a 29.5% of Tweets have scored a value of 0, followed by a 45% of postive scores (28% scored 1, 15% scored 2, 1.6% scored 3) and a 25% of negative result**

In [62]:
complete['sent_score'].value_counts(normalize=True)

 0   0.29524
 1   0.28333
 2   0.15238
-1   0.15238
-2   0.06905
-3   0.02619
 3   0.01667
-4   0.00476
Name: sent_score, dtype: float64

**Making a function which will summarize the results of the sentiment analysis and display them in a column simply saying if the Tweet is supposed to be a positive or a negative one. task 1 completed** 

In [63]:
def recategorize(category):
    if category == 0:
        return 'neutral'
    if category == 1:
        return 'positive'
    if category == 2:
        return 'positive'
    if category == 3:
        return 'positive'
    if category == 4:
        return 'positive'
    if category == -1:
        return 'negative'
    if category == -2:
        return 'negative'
    if category == -3:
        return 'negative'
    if category == -4:
        return 'negative'
    if category == -5:
        return 'negative'
    else:
        return 'Other'

In [64]:
complete['sentiment'] = complete['sent_score'].apply(recategorize)

**The algorithm on the sentiment score seems to perform well! The first tweet of the sample shows a positive text and it is reported as a positive sentiment score. The second tweet of the sample shows instead a quite negative text and the algorithm reported id as negative. It does look quite relaible and I do agree with all of the judgement it made in the sample**

In [65]:
complete.head()

id    screen_name                created_at  \
0  1438236436084391936    CitizenWonk 2021-09-15 20:21:02+00:00   
1  1438236402114772995  mark_trudgeon 2021-09-15 20:20:54+00:00   
2  1438236275283214343     AspieandMe 2021-09-15 20:20:23+00:00   
3  1438235849058045956      gazetka75 2021-09-15 20:18:42+00:00   
4  1438235834415714308      arynpekik 2021-09-15 20:18:38+00:00   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/android" ...   
2  <a href="https://mobile.twitter.com" rel="nofo...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/android" ...   

                                           full_text  retweet_count  \
0  RT @PBIcanada: We congratulate @GreenpeaceCA o...              7   
1  RT @cadebe_: @GJGamble @BorisJohnson @trussliz...              1   
2  RT @GreenpeaceCA: I scream, you scream, WE ALL...              2   
3  RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...             16   
4  RT @zoev213: Happy 50th @greenpeaceusa Join me...              2   

   favorite_count  followers_count_no_na  friendsCount_no_na  positive  \
0               0            32259.00000         27188.00000         3   
1               0              340.00000          1516.00000         1   
2               0              387.00000           672.00000         2   
3               0                1.00000            35.00000         2   
4               0              446.00000           443.00000         3   

   negative  neutral  sent_score sentiment  
0        -1        1           2  positive  
1        -1        0           0   neutral  
2        -5       -1          -3  negative  
3        -1        1           1  positive  
4        -1        1           2  positive

**As we can see the average of the Retweets and  Favorites is higher on Positive Tweets rather than Negative or Neutral Tweets and this already partially ansered my Research Question: the positive sentiment expressed in a post (Tweet in this case) positively influence user engagment**

In [66]:
complete.groupby('sentiment')[['retweet_count','favorite_count']].describe().transpose()

sentiment             negative   neutral  positive
retweet_count  count 106.00000 124.00000 190.00000
               mean   11.41509  13.32258  38.11579
               std    34.58898  36.35161  73.60518
               min     0.00000   0.00000   0.00000
               25%     0.00000   0.00000   1.00000
               50%     0.00000   2.00000  16.00000
               75%    15.00000  14.00000  62.75000
               max   268.00000 333.00000 648.00000
favorite_count count 106.00000 124.00000 190.00000
               mean    0.48113   0.55645   3.54737
               std     2.01533   3.63207  43.67757
               min     0.00000   0.00000   0.00000
               25%     0.00000   0.00000   0.00000
               50%     0.00000   0.00000   0.00000
               75%     0.00000   0.00000   0.00000
               max    20.00000  38.00000 602.00000

**The avergae Sentiment Analysis score is slightly above the zero so the Tweets extracted with Twarc were mostly positive Tweets. Task 2 completed**

In [67]:
complete['sent_score'].describe()

count   420.00000
mean      0.25000
std       1.31337
min      -4.00000
25%      -1.00000
50%       0.00000
75%       1.00000
max       3.00000
Name: sent_score, dtype: float64

**Making a sample of the Dataframe as requested by the project**

In [68]:
sample = complete.sample(15)

**Saving the Dataframe**

In [69]:
complete.to_pickle('COMPLETE.pkl')

**Importing Counter library just for curiosity in order to see which were the most common words used in ther Tweets**

In [70]:
import re
from collections import Counter

In [71]:
texts = complete['full_text'].values.tolist()

**Writing a function able to count the words present among the Tweets**

In [72]:
total_words = Counter()
for text in texts:
    # making text lower case
    text = text.lower()
    # removing URLS 
    text = text.split(' ')
    newtext = []
    for item in text:
        if 'http' not in item:
            newtext.append(item)
    
    newtext = ' '.join(newtext)
        
    # splitting the text in words (tokens)
    tokens = re.findall(r"[\w']+|[.,!?;$@]", newtext)
    for token in tokens:
        total_words[token] += 1
    

**Asking Python to show me the 100 most common words used in the Tweets**

In [73]:
total_words.most_common(100)

[('@', 1127),
 (',', 415),
 ('.', 360),
 ('the', 295),
 ('greenpeace', 287),
 ('rt', 255),
 ('to', 248),
 ('a', 162),
 ('of', 160),
 ('and', 148),
 ('!', 141),
 ('is', 140),
 ('in', 134),
 ('you', 131),
 ('for', 109),
 ('we', 86),
 ('today', 72),
 ('this', 67),
 ('happy', 61),
 ('our', 60),
 ('are', 58),
 ('on', 56),
 ('it', 55),
 (';', 54),
 ('years', 54),
 ('like', 52),
 ('i', 51),
 ('from', 51),
 ('all', 49),
 ('birthday', 48),
 ('?', 46),
 ('that', 45),
 ('action', 43),
 ('day', 42),
 ('can', 41),
 ('50th', 39),
 ('s', 38),
 ('green', 34),
 ('gt', 34),
 ('be', 32),
 ('greenpeaceusa', 32),
 ('thank', 31),
 ('have', 30),
 ('more', 30),
 ('greenpeaceca', 29),
 ('need', 29),
 ('climate', 29),
 ('so', 29),
 ('anniversary', 28),
 ('luisamneubauer', 28),
 ('seashepherd', 27),
 ('as', 27),
 ('50', 27),
 ('borisjohnson', 26),
 ('chrisgpackham', 26),
 ('seasaver', 26),
 ('carolinelucas', 26),
 ('change', 26),
 ('what', 26),
 ('has', 26),
 ('gjgamble', 25),
 ('trussliz', 25),
 ('environmental